### DISCUSSION TITLE
ONNX Runtime v1.16.0

### DISCUSSION THREAD

**Discussion by faxu** [microsoft/onnxruntime#17636] [General]
# General
* Support for serialization of models >=2GB 

# APIs
* New session option to disable default CPU EP fallback `session.disable_cpu_ep_fallback`
* Java
  * Support for fp16 and bf16 tensors as inputs and outputs, along with utilities to convert between these and fp32 data. On JDK 20 and newer the fp16 conversion methods use the JDK's Float.float16ToFloat and Float.floatToFloat16 methods which can be hardware accelerated and vectorized on some platforms.
  * Support for external initializers so that large models that can be instantiated without filesystem access
* C#
  * Expose OrtValue API as the new preferred API to run inference in C#. This reduces garbage and exposes direct native memory access via Slice like interfaces.
  * Make Float16 and BFloat16 full featured fp16 interfaces that support conversion and expose floating properties (e.g. IsNaN, IsInfinity etc)

# Performance
* Improve LLM quantization accuracy with smoothquant
* Support 4-bit quantization on CPU
* Optimize BeamScore to improve BeamSearch performance
* Add FlashAttention v2 support for Attention, MultiHeadAttention and PackedMultiHeadAttention ops 

# Execution Providers
* CUDA EP
  * Initial fp8 support (QDQ, Cast, MatMul) 
  * Relax CUDA Graph constraints to allow more models to utilize
  * Allow CUDA allocator to be registered with ONNX Runtime externally
* TensorRT EP
  * CUDA Graph support
  * Support user provided cuda compute stream
  * Misc bug fixes and improvements
* OpenVINO EP
  * Support OpenVINO 2023.1
* QNN EP
  * Enable context binary cache to reduce initialization time
  * Support QNN 2.12
  * Support for resize with asymmetric transformation mode on HTP backend
  * Ops support: Equal, Less, LessOrEqual, Greater, GreaterOrEqual, LayerNorm, Asin, Sign, DepthToSpace, SpaceToDepth
  * Support 1D Conv/ConvTranspose
  * Misc bug fixes and improvements

# Mobile
  * Initial support for [Azure EP](https://onnxruntime.ai/docs/execution-providers/Azure-ExecutionProvider.html)
  * Dynamic shape support for CoreML
  * Improve React Native performance with JSI
  * Mobile support for CLIPImageProcessor pre-processing and CLIP scenario
  * Swift Package Manager support for ONNX Runtime inference and ONNX Runtime extensions via [onnxruntime-swift-package-manager](https://github.com/microsoft/onnxruntime-swift-package-manager)

# Web
  * webgpu ops coverage improvements (SAM, T5, Whisper)
  * webnn ops coverage improvements (SAM, Stable Diffusion)
  * Stability/usability improvements for webgpu 

# Large model training
* ORTModule + OpenAI Triton Integration now available. [See details here](https://github.com/microsoft/onnxruntime/blob/main/docs/ORTModule_Training_Guidelines.md#6-use-openai-triton-to-compute-onnx-sub-graph)
* [Label Sparsity compute optimization](https://github.com/microsoft/onnxruntime/blob/main/docs/ORTModule_Training_Guidelines.md#ortmodule_enable_compute_optimizer) support complete and enabled by default starting release 1.16
* **New experimental** embedding [sparsity related optimizations](https://github.com/microsoft/onnxruntime/blob/main/docs/ORTModule_Training_Guidelines.md#ortmodule_enable_embedding_sparse_optimizer) available (disabled by default).
  * Improves training performance of Roberta in Transformers by 20-30%
* Other compute optimizations like Gather/Slice/Reshape upstream support enabled.
* Optimizations for [LLaMAv2 (~10% acceleration)](https://github.com/huggingface/optimum/tree/main/examples/onnxruntime/training/text-classification#text-classification) and OpenAI Whisper
* Improvements to logging and metrics (initialization overhead, memory usage, statistics convergence tool, etc) system improvements.
* PythonOp enhancement: bool and tuple[bool] constants, materialize grads, empty inputs, save in context, customized shape inference, use full qualified name for export.
* SCELossInternal/SCELossGradInternal CUDA kernels can handle elements more than std::numeric_limits<int32_t>::max.
* Improvements to LayerNorm fusion
* [Model cache](https://github.com/microsoft/onnxruntime/blob/main/docs/ORTModule_Training_Guidelines.md#ortmodule_cache_dir) for exported onnx model is introduced to avoid repeatedly exporting a model that is not changed across.
		
# On-Device Training
* iOS support available starting this release
* Minimal build now available for On-Device Training. Basic binary size ~1.5 MB
* ORT-Extensions custom op support enabled through onnxblock for on-device training scenarios

# ORT Extensions
This ORT release is accompanied by updates to [onnxruntime-extensions](https://github.com/microsoft/onnxruntime-extensions/). Features include:
* New Python API gen_processing_models to export ONNX data processing model from Huggingface Tokenizers such as LLaMA , CLIP, XLM-Roberta, Falcon, BERT, etc.
* New TrieTokenizer operator for RWKV-like LLM models, and other tokenizer operator enhancements.
* New operators for Azure EP compatibility: AzureAudioToText, AzureTextToText, AzureTritonInvoker for Python and NuGet packages.
* Processing operators have been migrated to the new [Lite Custom Op API](https://github.com/microsoft/onnxruntime/blob/gh-pages/docs/reference/operators/add-custom-op.md#define-and-register-a-custom-operator)
---
# Known Issues
* ORT CPU Python package requires execution provider to be explicitly provided. See #17631. Fix is in progress to be patched.
---
# Contributions
Contributors to ONNX Runtime include members across teams at Microsoft, along with our community members:
[fs-eire](https://github.com/fs-eire), [edgchen1](https://github.com/edgchen1), [snnn](https://github.com/snnn), [pengwa](https://github.com/pengwa), [mszhanyi](https://github.com/mszhanyi), [PeixuanZuo](https://github.com/PeixuanZuo), [tianleiwu](https://github.com/tianleiwu), [adrianlizarraga](https://github.com/adrianlizarraga), [baijumeswani](https://github.com/baijumeswani), [cloudhan](https://github.com/cloudhan), [satyajandhyala](https://github.com/satyajandhyala), [yuslepukhin](https://github.com/yuslepukhin), [RandyShuai](https://github.com/RandyShuai), [RandySheriffH](https://github.com/RandySheriffH), [skottmckay](https://github.com/skottmckay), [Honry](https://github.com/Honry), [dependabot[bot]](https://github.com/dependabot[bot]), [HectorSVC](https://github.com/HectorSVC), [jchen351](https://github.com/jchen351), [chilo-ms](https://github.com/chilo-ms), [YUNQIUGUO](https://github.com/YUNQIUGUO), [justinchuby](https://github.com/justinchuby), [PatriceVignola](https://github.com/PatriceVignola), [guschmue](https://github.com/guschmue), [yf711](https://github.com/yf711), [Craigacp](https://github.com/Craigacp), [smk2007](https://github.com/smk2007), [RyanUnderhill](https://github.com/RyanUnderhill), [jslhcl](https://github.com/jslhcl), [wschin](https://github.com/wschin), [kunal-vaishnavi](https://github.com/kunal-vaishnavi), [mindest](https://github.com/mindest), [xadupre](https://github.com/xadupre), [fdwr](https://github.com/fdwr), [hariharans29](https://github.com/hariharans29), [AdamLouly](https://github.com/AdamLouly), [wejoncy](https://github.com/wejoncy), [chenfucn](https://github.com/chenfucn), [pranavsharma](https://github.com/pranavsharma), [yufenglee](https://github.com/yufenglee), [zhijxu-MS](https://github.com/zhijxu-MS), [jeffdaily](https://github.com/jeffdaily), [natke](https://github.com/natke), [jeffbloo](https://github.com/jeffbloo), [liqunfu](https://github.com/liqunfu), [wangyems](https://github.com/wangyems), [er3x3](https://github.com/er3x3), [nums11](https://github.com/nums11), [yihonglyu](https://github.com/yihonglyu), [sumitsays](https://github.com/sumitsays), [zhanghuanrong](https://github.com/zhanghuanrong), [askhade](https://github.com/askhade), [wenbingl](https://github.com/wenbingl), [jingyanwangms](https://github.com/jingyanwangms), [ashari4](https://github.com/ashari4), [gramalingam](https://github.com/gramalingam), [georgen117](https://github.com/georgen117), [sfatimar](https://github.com/sfatimar), [BowenBao](https://github.com/BowenBao), [hanbitmyths](https://github.com/hanbitmyths), [stevenlix](https://github.com/stevenlix), [jywu-msft](https://github.com/jywu-msft)

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/onnxruntime/releases/tag/v1.16.0'>ONNX Runtime v1.16.0</a>.</em>

**Comment by PenutChen** [comment]
Is there any documents about 4-bit quantization?

**Comment by csukuangfj** [comment]
Please also add 
https://github.com/microsoft/onnxruntime/issues/17645
to known issues.

**Comment by qinhj** [comment]
It seems that the "onnxruntime_float16.h" is missing (e.g. onnxruntime-linux-x64-1.16.0.tgz), which is required(i.e. #include) in "onnxruntime-linux-x64-1.16.0.tgz/include/onnxruntime_cxx_api.h"

**Reply by csukuangfj** [reply]
Yes, you are right. Please see a similar issue at
https://github.com/microsoft/onnxruntime/issues/17645

**Reply by csukuangfj** [reply]
You are not alone. More and more people will have the same issue.

**Comment by pfeatherstone** [comment]
Hopefully there will be a 1.16.1 release soon.

**Reply by faxu** [reply]
Yes, this is in progress and will fix known issues. Thanks for your patience.

**Comment by greyovo** [comment]
Is there any detail information about _"Mobile support for CLIPImageProcessor pre-processing and CLIP scenario"_ ?
